### DISCUSSION TITLE
Implement RBAC

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#60933] [Help] [answered]
### Summary

I am wondering how to create a role-based access control (RBAC) system with Next.js. For instance, [Stripe has an RBAC system](https://strapi.io/features/custom-roles-and-permissions) with roles such as admin, moderator, staff, writer, and viewer. Each role has different permissions, for example, only admins can create posts, but viewers and commuters cannot. Similarly, [MedusaJS also has an RBAC system](https://docs.medusajs.com/modules/users/backend/rbac#create-api-routes-and-services,) but it uses Express.js as its backend. I need to implement something similar with Next.js. I don't want to add a field like "is_admin" to the database, but I also need permission to check and conditionally render components like buttons. Thank you in advance.

**Comment by ghost** [comment, marked as answer]
I got a good answer on Reddit regarding how to implement role-based access control with Next.js. The answer suggested using a library called CASl. I checked out the documentation, and it looks like it requires some programming https://casl.js.org/.

Reddit post - https://www.reddit.com/r/nextjs/comments/19c42ba/implement_rbac/

**Comment by gemanor** [comment]
In general, Next.js's fullstack nature makes it very friendly to implement RBAC (and authorization) for both the frontend and the backend.

In the backend, the best practice is to decouple the policy configuration (what admin can see, what paid user can click) from the code to an external policy-based authorization service. By decoupling this logic from the application code you're getting the ability to change it without changes in the application code, and also to scale for more [fine-grained authorization](https://www.permit.io/blog/what-is-fine-grained-authorization-fga) models when you need them. Externalizing authorization can be done by using open-source services such as [OPAL](https://github.com/permitio/opal), [AWS Cedar](https://www.cedarpolicy.com/), or [OpenFGA](https://openfga.dev/). You can read more about Next.js and RBAC using externalizing authorization here: https://www.permit.io/blog/how-to-add-rbac-in-nextjs

In the frontend, the best practice is to not enforce permissions (as people can change the state in the browser), but toggle the features/components per the permissions you enforce in the backend. You can use a library like CASL to enforce them. Here is an instruction of using CASL and Next.js: https://permit.io/blog/frontend-authorization-with-nextjs-and-casl-tutorial
